pallets/creditcoin/src/tests.rs

```diff
@@ -815,3 +1183,388 @@ fn add_authority_works_for_root() {
+		// erroring out when checking for their existence
+		let ask_order_id = AskOrderId::new::<Test>(test_info.expiration_block, &test_info.ask_guid);
+		let bid_order_id = BidOrderId::new::<Test>(test_info.expiration_block, &test_info.bid_guid);
+		let offer_id =
+			OfferId::new::<Test>(test_info.expiration_block, &ask_order_id, &bid_order_id);
+		let deal_order_id = DealOrderId::new::<Test>(test_info.expiration_block, &offer_id);
+
+		let deal_order = DealOrder {
+			blockchain: test_info.blockchain,
+			offer_id,
+			lender_address_id: test_info.lender.address_id.clone(),
+			borrower_address_id: test_info.borrower.address_id.clone(),
+			terms: test_info.loan_terms.clone(),
+			expiration_block: test_info.expiration_block,
+			timestamp: Creditcoin::timestamp(),
+			borrower: test_info.borrower.account_id,
+			funding_transfer_id: None,
+			lock: None,
+			repayment_transfer_id: None,
+		};
+
+		// insert this DealOrder into storage which will use-up the ID
+		// register_deal_order() will reconstruct the same ID later
+		crate::DealOrders::<Test>::insert_id(deal_order_id, deal_order);
+
+		let message = Creditcoin::register_deal_order_message(
+			test_info.expiration_block,
+			&test_info.ask_guid,
+			&test_info.bid_guid,
+		);
+		let signature = key_pair.sign(&message);
+
+		assert_noop!(
+			Creditcoin::register_deal_order(
+				Origin::signed(test_info.lender.account_id),
+				test_info.lender.address_id,
+				test_info.borrower.address_id,
+				test_info.loan_terms,
+				test_info.expiration_block,
+				test_info.ask_guid,
+				test_info.bid_guid,
+				key_pair.public(),
+				signature,
+			),
+			crate::Error::<Test>::DuplicateDealOrder
+		);
+	});
+}
+
+#[test]
+fn register_deal_order_should_succeed() {
+	ExtBuilder::default().build_and_execute(|| {
+		System::set_block_number(1);
+
+		let (key_pair, _) = sp_core::ecdsa::Pair::generate();
+		let test_info = TestInfo {
+			borrower: RegisteredAddress::from_pubkey(key_pair.public(), Blockchain::Rinkeby),
+			..TestInfo::new_defaults()
+		};
+
+		let message = Creditcoin::register_deal_order_message(
+			test_info.expiration_block,
+			&test_info.ask_guid,
+			&test_info.bid_guid,
+		);
+		let signature = key_pair.sign(&message);
+
+		assert_ok!(Creditcoin::register_deal_order(
+			Origin::signed(test_info.lender.account_id),
+			test_info.lender.address_id,
+			test_info.borrower.address_id,
+			test_info.loan_terms,
+			test_info.expiration_block,
+			test_info.ask_guid,
+			test_info.bid_guid,
+			key_pair.public(),
+			signature,
+		));
+
+		let event = <frame_system::Pallet<Test>>::events()
+			.pop()
+			.expect("Expected at least one EventRecord to be found")
+			.event;
+
+		let tname = type_of(event);
+		log::debug!("**** DEBUG: type_name={:?}", tname);
+
+		// TODO: assert on event types b/c the actual values are hidden inside the
+		// implementation of register_deal_order()
+		//		assert_type_eq!(event, crate::Event::<Test>::AskOrderAdded);
```

Comment:
@nathanwhit do you have any idea how we can assert that the right events have been fired? The [available examples](https://docs.substrate.io/v3/runtime/testing/) are asserting events for equality and that requires all of the arguments, which are not exposed outside of `register_deal_order()`. 

I wanted to see if we can assert the data types directly but given that Rust is a statically and strongly typed language all hints I could find discourage the programmer to actually do this so I can't find something suitable which could work for the sake of making a few assertions.

Reply:
`crate::Event::<Test>::AskOrderAdded` is an enum variant, not a type, so we really just want to check if `event` matches that variant. So you can just assert that with a wildcard pattern `..` to indicate you don't care about the fields:

```rust
assert!(matches!(event, crate::mock::Event::Creditcoin(crate::Event::AskOrderAdded(..)));
```